tests/cpgrid/LgrChecks.hpp

```diff
@@ -0,0 +1,466 @@
+        }
+    }
+    BOOST_CHECK_CLOSE(referenceElemOneParent_volume_it, 1, 1e-13);
+
+    for (int c = 0; c < 3; ++c)
+    {
+        referenceElem_entity_center_it[c]/= total_children;
+        BOOST_CHECK_CLOSE(referenceElem_entity_center_it[c], .5, 1e-13);
+    }
+}
+
+void Opm::checkGlobalCellBounds(const Dune::CpGrid& grid,
+                                const std::vector<std::shared_ptr<Dune::cpgrid::CpGridData>>& data)
+{
+    const int maxLevel = grid.maxLevel();
+    for (int level = 0; level <= maxLevel; ++level) {
+        if (!(data[level] -> globalCell().empty())) { // If level>0, in some processes, LGR might be empty.
+            Opm::checkGlobalCellBounds(data[level]->globalCell(), data[level]->logicalCartesianSize());
+        }
+    }
+    Opm::checkGlobalCellBounds(data.back()->globalCell(), data.back()->logicalCartesianSize());
+    Opm::checkEqMinMaxGlobalCellLevelZeroAndLeaf(data.front()->globalCell(), data.back()->globalCell());
+}
+
+void Opm::checkGlobalCellBounds(const std::vector<int>& globalCell,
+                                const std::array<int, 3>& logicalCartesianSize)
+{
+    const auto [itMin, itMax] = std::minmax_element(globalCell.begin(), globalCell.end());
+    BOOST_CHECK( *itMin >= 0);
+    //Note:  An LGR can have cells distributed across different processes, so the minimum cell global id may not be zero in all processes.
+
+    const auto maxCartesianIdxLevel = logicalCartesianSize[0]*logicalCartesianSize[1]*logicalCartesianSize[2];
+    BOOST_CHECK( *itMax < maxCartesianIdxLevel);
+}
+
+void Opm::checkEqMinMaxGlobalCellLevelZeroAndLeaf(const std::vector<int>& globalCell_l0,
+                                                  const std::vector<int>& globalCell_leaf)
+{
+    const auto [itMinL0, itMaxL0] = std::minmax_element(globalCell_l0.begin(), globalCell_l0.end());
+    const auto [itMinLeaf, itMaxLeaf] = std::minmax_element(globalCell_leaf.begin(), globalCell_leaf.end());
+    BOOST_CHECK_EQUAL( *itMinL0, *itMinLeaf);
+    BOOST_CHECK_EQUAL( *itMaxL0, *itMaxLeaf);
+}
+
+void Opm::checkFatherAndSiblings(const Dune::cpgrid::Entity<0>& element,
+                                 int level,
+                                 double expected_total_children,
+                                 const Dune::CpGrid& grid)
+{
+    BOOST_CHECK( element.father().level() == 0); // If there is no nested refinement
+    BOOST_CHECK( element.father().isLeaf() == false); // Father vanished during refinement.
+    BOOST_CHECK( element.father() == element.getOrigin() );
+
+    auto itFather = element.father().hbegin(grid.maxLevel());
+    const auto endItFather = element.father().hend(grid.maxLevel());
+    // If itFather != endItFather and !element.father().isLeaf() (if dristibuted_data_ is empty).
+    BOOST_CHECK( itFather != endItFather );
+    Opm::checkAverageCenterAndVolume(itFather, endItFather, expected_total_children);
+
+    const auto [child_level, siblings_list] =
```

Comment:
Do we need copies?

Reply:
same here, no copies needed. I'll correct that